src/content/blog/fastify.mdx

```diff
@@ -0,0 +1,179 @@
+We've gone from CommonJS to ESM, from browserify to webpack to vite,
+from mocha to vitest, from the request module to using web-standard
+fetch everywhere. The constant reinvention can
+be exciting or annoying from one day to the next.
+
+### The success of Express
+
+In the midst of all this churn, [express][express] has persevered.
+Express is a web framework for Node.js, introduced
+by [TJ Holowaychuk][tj] in 2009. Like a lot
+of early JavaScript modules and many of TJ's projects, it took inspiration
+from the land of Ruby - in particular, [Sinatra][sinatra], which
+was established in 2007.
+
+```js title="express-example.js"
+// This is a tiny example of using
+// express. Notice how there's almost no boilerplate
+// involved in creating a tiny web server.
+const app = require('express')();
+app.get('/', (req, res) => {
+  res.send('Hello World!');
+});
+app.listen(3000);
+```
+
+Express is a beautifully light abstraction: a
+'hello world' example is barely over 10 lines. Compared to the boilerplate
+required to do the same in Ruby on Rails or a Java server at the time, it
+was a breath of fresh air.
+
+<NpmChart />
+
+Express became the default for new JavaScript projects, and has stayed
+there. In a typical week, express is downloaded from NPM 30 million
+times. It's a roaring success in every way.
+
+### A need for types
+
+Express is an _expressive_ framework: like a classic Ruby or
+JavaScript framework, it lets you color outside the lines by using
+loosely-defined objects and types. Query string parameters in express
+become [freeform objects](https://expressjs.com/en/4x/api.html#req.query).
+The same goes for the bodies of requests, and the kinds of responses
+you can send: it's all a bit ad-hoc by default.
+
+But the vibes have shifted: we write [TypeScript][ts] now,
+and everyone's excited about validator modules like [Zod](https://zod.dev/)
+and [TypeBox](https://github.com/sinclairzx81/typebox). Interfaces
+are less ad-hoc and more declared, tested, and strictly typed.
+
+Modern tooling is written to take full advantage of types
+and schemas. With [tRPC](https://trpc.io/), we're writing backend
+endpoints and calling them from auto-generated, auto-typed functions
+on the frontend. Other backends are embracing [OpenAPI][openapi]
+(née Swagger) to generate clients, test suites, and more.
+End-to-end typesafety is in vogue, and it is good.
+
+###  Issues with express
+
+Val Town started off with Express. It has served us pretty well.
+
+But the papercuts started to accumulate.
+When we started creating  [a public API with an OpenAPI spec][rfc],
+we immediately started to get a bad feeling about using express
+as we were doing before. The OpenAPI specification was making
+strong promises about the kinds of requests we accepted and responses
+we'd return. But the specification was just a dead document,
+a YAML file - it wasn't part our codebase, or our test suite. So it
+wasn't enforced, and what isn't enforced isn't true.
```

Comment:
You could also mention:

- our yaml file was never valid
- we wanted an sdk and that was hard (and you could link to https://macwright.com/2024/06/21/apis-and-applications)
- that openapi's are all the sudden more relevant in the age of llm's reading them for some reason